modules/gapi/include/opencv2/gapi/rmat.hpp

```diff
@@ -0,0 +1,127 @@
+    int depth() const { return m_desc.depth; }
+    int chan() const { return m_desc.chan; }
+    size_t elemSize() const { return CV_ELEM_SIZE(type()); }
+
+    template<typename T = uchar> T* ptr(int y = 0, int x = 0) {
+        return reinterpret_cast<T*>(m_data + m_step*y + x*CV_ELEM_SIZE(type()));
+    }
+    template<typename T = uchar> const T* ptr(int y = 0, int x = 0) const {
+        return reinterpret_cast<const T*>(m_data + m_step*y + x*CV_ELEM_SIZE(type()));
+    }
+    size_t step() const { return m_step; }
+
+private:
+    GMatDesc m_desc;
+    uchar* m_data = nullptr;
+    size_t m_step = 0u;
+    DestroyCallback m_cb = nullptr;
+};
+
+// "Remote Mat", a general class which provides an abstraction layer over the data
+// storage and placement (host, remote device etc) and allows to access this data.
+//
+// The device specific implementation is hidden in the RMat::Adapter class
+//
+// The basic flow is the following:
+// * Backend which is aware of the remote device:
+//   - Implements own AdapterT class which is derived from RMat::Adapter
+//   - Wraps device memory into RMat via make_rmat utility function:
+//         cv::RMat rmat = cv::make_rmat<AdapterT>(args);
+//
+// * End user:
+//   - Writes the code which works with RMats without any knowledge of the remote device:
+//     void func(const cv::RMat& in_rmat, cv::RMat& out_rmat) {
+//         // Fetch input data from the device, get mapped memory for output
+//         cv::MatView  in_view =  in_rmat.access(Access::R);
+//         cv::MatView out_view = out_rmat.access(Access::W);
+//         performCalculations(in_view, out_view);
+//         // data from out_view is transferred to the device when out_view is destroyed
+//     }
+class RMat
+{
+public:
+    enum class Access { R, W };
+    class Adapter
+    {
+    public:
+        virtual ~Adapter() = default;
+        virtual GMatDesc desc() const = 0;
+        // Implemetation is responsible for setting the appopriate callback to
+        // the view when accessed for writing, to ensure that the data from the view
+        // is tranferred to the device when the view is destroyed
+        virtual cv::MatView access(Access) const = 0;
+    };
+    using AdapterP = std::shared_ptr<Adapter>;
+
+    RMat() = default;
+    // FIXME? make private? (make make_rmat a friend?)
+    RMat(AdapterP&& a) : m_adapter(std::move(a)) {}
+    GMatDesc desc() const { return m_adapter->desc(); }
+
+    // Note: When accessed for write there is no guarantee that returned view
+    // will contain actual snapshot of the mapped device memory
+    // (no guarantee that fetch from a device is performed). The only
+    // guaranty is that when the view is destroyed, its data will be
+    // transferred to the device
+    cv::MatView access(Access a) const { return m_adapter->access(a); }
+
+    // Check that underlying RMat adapter is of the partucular type
+    template<typename T> bool is() const
```

Comment:
Can we have a better naming than is/as? It's at least confusing compared to what Blobs in IE used to have (not sure if it changed). IMO is()/as() implies a check on the dynamic type of RMat itself, not on the adapter that it stores.

Maybe holds_adapter() / get_adapter() (or something less verbose)?

Reply:
Renamed